working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,937 @@
+of an operator that takes at least one argument), it is allowed to pass an
+actual argument list, and the static analysis of the actual arguments
+proceeds as with other function calls, using a signature where the formal
+type parameters of `E` are replaced by `T1, .. Tk`. The type of the entire
+member access is the return type of said member if it is a member
+invocation, and the function type of the method if it is an extension
+member tear-off, again substituting `T1, .. Tk` for the formal type
+parameters.
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type.
+
+*For example, it can occur as the declared type of a variable or parameter,
+as the return type of a function or getter, as a type argument in a type,
+as the on-type of an extension, as the type in the `onPart` of a
+try/catch statement, or in a type test `o is E` or a type cast `o as E`, or
+as the body of a type alias. It is also allowed to create a new instance
+where one or more extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+A compile-time error occurs if an extension type is used as a
+superinterface of a class or mixin, or if an extension type is used to
+derive a mixin.
+
+*So `class C extends E1 with E2 implements E3 {}` has three errors if `E1`,
+`E2`, and `E3` are extension types, and `mixin M on E1 implements E2 {}`
+has two errors.*
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>invokeExtensionMethod(Ext<S<sub>1</sub>, .. S<sub>k</sub>>, e).m(args)</code>
+and similarly for instance getters and operators. This rule also applies
+when a member access implicitly has the receiver `this`, and the static
+type of `this` is an extension type (*which can only occur in an extension
+type member declaration where the on-type is itself an extension type*).
+
+*That is, when the type of an expression is an extension type, all method
+invocations on that expression will invoke an extension method declared by
+that extension, and similarly for other member accesses. In particular, we
+cannot invoke an instance member when the receiver type is an extension
```

Comment:
I'm quite confused again about what semantics you intend for `this` calls again.  I think this text is self-contradictory in a number of places: Specifically: "when the type of an expression is an extension type, all method
invocations on that expression will invoke an extension method declared by
that extension, and similarly for other member accesses." from above is in direct contradiction with "Similarly, extension methods of _E_ invoked in the body of _E_ are
subject to the same treatment as previously, which means that extension
methods of the enclosing extension type can be invoked implicitly, and
extension methods are given higher priority than instance methods on `this`
when `this` is implicit.*".